src/OpenLoco/src/World/TownManager.cpp

```diff
@@ -29,12 +34,140 @@ namespace OpenLoco::TownManager
+        {
+            char buffer[256]{};
+            auto rand = town->prng.randNext();
+            auto testsToRun = townNameFromNamesObject(rand, buffer);
+
+            if (strlen(buffer) == 0)
+            {
+                continue;
+            }
+
+            if (strlen(buffer) > StringManager::kUserStringSize)
+            {
+                continue;
+            }
+
+            if (Gfx::TextRenderer::getStringWidth(Gfx::Font::medium_bold, buffer) > 200)
+            {
+                continue;
+            }
+
+            // clang-format off
+            if (testsToRun & 1)
+            {
+                // Check that the town is adjacent to a large amount of water tiles on at least one side.
+                auto pos = Pos2(town->x, town->y);
+                if (!(TileManager::countSurroundingWaterTiles(pos + Pos2(6 * kTileSize, 0)) > 65 ||
+                    TileManager::countSurroundingWaterTiles(pos + Pos2(0, 6 * kTileSize)) > 65 ||
+                    TileManager::countSurroundingWaterTiles(pos + Pos2(0 - 6 * kTileSize, 0)) > 65 ||
+                    TileManager::countSurroundingWaterTiles(pos + Pos2(0, 0 - 6 * kTileSize)) > 65))
+                {
+                    continue;
+                }
+            }
+
+            if (testsToRun & 2)
+            {
+                auto pos = Pos2(town->x + kTileSize / 2, town->y + kTileSize / 2);
+                auto height = TileManager::getHeight(pos);
+                if (height.landHeight < 192)
+                {
+                    continue;
+                }
+            }
+
+            if (testsToRun & 4)
+            {
+                // Check that the town is adjacent to a low amount of water tiles on at least one side.
+                auto pos = Pos2(town->x, town->y);
+                if (!(TileManager::countSurroundingWaterTiles(pos + Pos2(6 * kTileSize, 0)) > 15 ||
+                    TileManager::countSurroundingWaterTiles(pos + Pos2(0, 6 * kTileSize)) > 15 ||
+                    TileManager::countSurroundingWaterTiles(pos + Pos2(0 - 6 * kTileSize, 0)) > 15 ||
+                    TileManager::countSurroundingWaterTiles(pos + Pos2(0, 0 - 6 * kTileSize)) > 15))
+                {
+                    continue;
+                }
+            }
+            // clang-format on
+
+            bool nameInUse = false;
+            for (auto& candidateTown : towns())
+            {
+                // Ensure the town name doesn't exist yet
+                char candidateTownName[256]{};
+                StringManager::formatString(candidateTownName, candidateTown.name);
+
+                if (strcmp(buffer, candidateTownName) == 0)
+                {
+                    nameInUse = true;
+                    break;
+                }
+            }
+
+            if (nameInUse)
+                continue;
+
+            StringId newNameId = StringManager::userStringAllocate(buffer, 0);
```

Comment:
not related to this PR but `changeCompanyName` and `changeCompanyOwnerName` have called this function incorrectly they are meant to call it with cl set to non-zero which means name must be unique in the user string table